python_modules/dagster-externals/dagster_externals/_io/default.py

```diff
@@ -1,57 +1,65 @@
 import json
+from contextlib import contextmanager
+from typing import Any, Iterator, Mapping, Tuple
 
 from .._protocol import (
     DAGSTER_EXTERNALS_ENV_KEYS,
     ExternalExecutionContextData,
     ExternalExecutionMessage,
+    ExternalExecutionParams,
 )
 from .._util import DagsterExternalsError, param_from_env_var
-from .base import ExternalExecutionContextLoader, ExternalExecutionMessageWriter
+from .base import (
+    ExternalExecutionContextLoader,
+    ExternalExecutionMessageWriter,
+    ExternalExecutionParamLoader,
+)
 
 
-class ExternalExecutionFileContextLoader(ExternalExecutionContextLoader):
-    def load_context(self) -> ExternalExecutionContextData:
-        path = self._get_path_from_env()
-        with open(path, "r") as f:
-            return json.load(f)
+class PathMixin:
+    @contextmanager
+    def setup(self, params: ExternalExecutionParams) -> Iterator[None]:
+        self._validate_params(params)
+        self._path = params["path"]
+        yield
```

Comment:
Can we make this instead an abstract class method that ends up return an instance of class? I generally think these patterns where are an is constructed and then is setup/initialized with a non-ctor method is stateful and error-prone.

Reply:
>I generally think these patterns where are an is constructed and then is setup/initialized with a non-ctor method is stateful and error-prone.

I agree and have been mentally turning this over looking for a better solution.

>Can we make this instead an abstract class method that ends up return an instance of class?

So the fundamental issue here is that we have two sets of params (1) user-provided params on the ext side, which might be anything for a custom loader/writer. (2) params passed over the orch side which need to be extracted from the environment. Both sets of params need to be combined in the same loader or writer object.

The difficulty is that they come from different sources and are available at different times. (1) are specified by the user in the external script. Since (2) are passed by the orchestrator, they are available only at context initialization time.

I see 3 ways to handle this scenario:

- **A (status quo)**: User passes an instance to `init_dagster_externals`. Instance has been constructed with params (1). The instance itself is responsible for extracting (2) from the environment and modifying itself accordingly.
- **B (this PR)**: User passes an instance to `init_dagster_externals`. Instance has been constructed with params (1). `init_dagster_externals` handles extraction of (2) from env and calls `setup` with extracted (2) to mutate the object.
- **C**: Same as A except instead of mutating the object, `init_dagster_externals` copies it and applies params (2) (perhaps the constructor accepts them).
- **D**: User passes some intermediate entity containing the class and params (1) that is not an instance. When params (2) are available they are combined with params (1) and passed together to the constructor.

I like (D), which lambdas fall under. What do you think of either of these:

```
### "lambda solution"
init_dagster_externals(
    context_loader=lambda params: CustomContextLoader(foo=foo, bar=bar, **params)
)

### "partials solution"
init_dagster_externals(
    # `partial` will be defined on base classes and return a simple object having
    # the `setup` method that returns `CustomContextLoader`
    context_loader=CustomContextLoader.partial(foo=foo, bar=bar)
)
```

If we want this to be pluggable in an intuitive way, IMO users should be able to pass an instance of a `ContextLoader`/`MessageWriter` to `init_dagster_externals`. This gives them full control over the "first round of parameter binding" (i.e. they can structure the constructor and its args however they'd like). And if they pass an instance, then the second round (params passed over from orch side)